pkg/keyspace/tso_keyspace_group.go

```diff
@@ -49,12 +84,94 @@ func (m *GroupManager) Bootstrap() error {
 	if err != nil && err != ErrKeyspaceGroupExists {
 		return err
 	}
-
+	if m.client != nil {
+		m.nodesBalancer = balancer.GenByPolicy[string](m.policy)
+		resp, err := etcdutil.EtcdKVGet(m.client, m.tsoServiceKey, clientv3.WithRange(m.tsoServiceEndKey))
+		if err != nil {
+			return err
+		}
+		for _, item := range resp.Kvs {
+			m.nodesBalancer.Put(string(item.Value))
+		}
+		m.wg.Add(1)
+		go m.startWatchLoop(resp.Header.GetRevision())
+	}
 	return nil
 }
 
+// Close closes the manager.
+func (m *GroupManager) Close() {
+	m.cancel()
+	m.wg.Wait()
+}
+
+func (m *GroupManager) startWatchLoop(revision int64) {
+	defer m.wg.Done()
+	if err := m.watchServiceAddrs(revision); err != nil {
+		log.Error("watch service addresses failed", zap.Error(err))
+	}
+}
+
+func (m *GroupManager) watchServiceAddrs(revision int64) error {
+	ctx, cancel := context.WithCancel(m.ctx)
+	defer cancel()
+	for {
+		select {
+		case <-ctx.Done():
+			return nil
+		default:
+		}
+		watchChan := m.client.Watch(ctx, m.tsoServiceKey, clientv3.WithRange(m.tsoServiceEndKey), clientv3.WithRev(revision))
+		for {
+			select {
+			case <-ctx.Done():
+				return nil
+			case wresp := <-watchChan:
+				if wresp.CompactRevision != 0 {
+					log.Warn("required revision has been compacted, the watcher will watch again with the compact revision",
+						zap.Int64("required-revision", revision),
+						zap.Int64("compact-revision", wresp.CompactRevision))
+					revision = wresp.CompactRevision
+					break
+				}
+				if wresp.Err() != nil {
```

Comment:
my understanding is that wresp.Err() != nil means two things here:
1. it isn't equal to ctx being cancelled. So if we return here and exit watchLoop but the API service is still serving and try to recover from some bad thing happened, once it's recovered we have no loop to watch the primary address change.
2. It doesn't help to call m.client.Watch to retry because watch.go will retry Watch() as much as it can unless it thinks it doesn't help.

It seems that we can't exit the watch loop but wait a while then retry "watcher := clientv3.NewWatcher(m.client) ; then watch" endlessly until ctx is cancelled. + @rleungx to  provide possible comment on this.

Reply:
I will do a fix similiar with #6279